aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,341 @@
+	bucketIdx int,
+	bucket *flushBucket,
+) {
+	nowNanos := mgr.nowNanos()
+	flushInterval := bucket.interval
+	nextFlushNanos := nowNanos + int64(flushInterval)
+	if mgr.jitterEnabled {
+		jitterNanos := int64(mgr.rand.Int63n(int64(flushInterval)))
+		nextFlushNanos = nowNanos + jitterNanos
+	}
+	newFlushMetadata := flushMetadata{
+		timeNanos: nextFlushNanos,
+		bucketIdx: bucketIdx,
+	}
+	mgr.flushTimes.Push(newFlushMetadata)
+}
+
+func (mgr *leaderFlushManager) prepareFlushTimesWithLock(
+	buckets []*flushBucket,
+) schema.ShardSetFlushTimes {
+	proto := schema.ShardSetFlushTimes{
+		ByShard: make(map[uint32]*schema.ShardFlushTimes, defaultInitialFlushTimesCapacity),
+	}
+	for _, bucket := range buckets {
+		for _, flusher := range bucket.flushers {
+			shard := flusher.Shard()
+			shardFlushTimes, exists := proto.ByShard[shard]
+			if !exists {
+				shardFlushTimes = &schema.ShardFlushTimes{
+					ByResolution: make(map[int64]int64, 2),
+				}
+				proto.ByShard[shard] = shardFlushTimes
+			}
+			resolution := flusher.Resolution()
+			lastFlushedNanos := flusher.LastFlushedNanos()
+			shardFlushTimes.ByResolution[int64(resolution)] = lastFlushedNanos
+		}
+	}
+	return proto
+}
+
+func (mgr *leaderFlushManager) persistFlushTimes(flushTimes *schema.ShardSetFlushTimes) {
+	persistStart := mgr.nowFn()
+	_, err := mgr.flushTimesStore.Set(mgr.flushTimesKey, flushTimes)
+	mgr.metrics.flushTimesPersist.ReportSuccessOrError(err, mgr.nowFn().Sub(persistStart))
+}
+
+func (mgr *leaderFlushManager) nowNanos() int64 { return mgr.nowFn().UnixNano() }
+
+type leaderFlushTask struct {
+	mgr           *leaderFlushManager
+	shouldFlush   bool
+	duration      tally.Timer
+	flushers      []PeriodicFlusher
+	shouldPersist bool
+	flushTimes    *schema.ShardSetFlushTimes
+}
+
+func (t *leaderFlushTask) Run() {
+	mgr := t.mgr
+
+	// NB(xichen): we intentionally do not use the worker pool as persistence could be
+	// done asynchronously to avoid blocking the workers reserved for flushing.
+	if t.shouldPersist {
+		go mgr.persistFlushTimes(t.flushTimes)
+	}
+
+	if t.shouldFlush {
+		var wgWorkers sync.WaitGroup
+		start := mgr.nowFn()
+		for _, flusher := range t.flushers {
+			flusher := flusher
+			wgWorkers.Add(1)
+			mgr.workers.Go(func() {
+				flusher.Flush()
+				wgWorkers.Done()
+			})
+		}
+		wgWorkers.Wait()
+		t.duration.Record(mgr.nowFn().Sub(start))
+	}
+}
+
+// flushMetadata contains metadata information for a flush.
+type flushMetadata struct {
+	timeNanos int64
+	bucketIdx int
+}
+
+// flushMetadataHeap is a min heap for flush metadata where the metadata with the
+// earliest flush time will be at the top of the heap. Unlike the generic heap in
+// the container/heap package, pushing data to or popping data off of the heap doesn't
+// require conversion between flush metadata and interface{}, therefore avoiding the
+// memory and GC overhead due to the additional allocations.
+type flushMetadataHeap []flushMetadata
```

Comment:
What's the max size of the heap? Is this one of those cases like in M3DB where we spend time to order a heap size of 3?